File: Python_mp/swapview.py
#!/usr/bin/env python3

from __future__ import print_function
from __future__ import division

from concurrent.futures import ProcessPoolExecutor
import os

format = "%5s %9s %s"
totalFmt = "Total: %8s"

def filesize(size):
  units = 'KMGT'
  left = abs(size)
  unit = -1
  while left > 1100 and unit < 3:
    left /= 1024
    unit += 1
  if unit == -1:
    return '%dB' % size
  else:
    if size < 0:
      left = -left
    return '%.1f%siB' % (left, units[unit])

def getSwapFor(pid):
  try:
    comm = open('/proc/%s/cmdline' % pid).read()
    if comm and comm[-1] == '\x00':
      comm = comm[:-1]
    comm = comm.replace('\x00', ' ')
    with open('/proc/%s/smaps' % pid) as f:
      s = sum(int(l.split()[1]) for l in f if l.startswith('Swap:'))
    return pid, s * 1024, comm
  except (IOError, OSError):
    return pid, 0, ''

def getSwap():
  futures = []
  with ProcessPoolExecutor() as executor:
    for pid in os.listdir('/proc'):
      if pid.isdigit():
        future = executor.submit(getSwapFor, pid)
        futures.append(future)

  ret = []
  for future in futures:
    s = future.result()
    if s[1] > 0:
      ret.append(s)
  ret.sort(key=lambda x: x[1])
  return ret

def main():
  results = getSwap()
  print(format % ('PID', 'SWAP', 'COMMAND'))
  for pid, swap, comm in results:
    print(format % (pid, filesize(swap), comm))
  t = filesize(sum(x[1] for x in results))
  print(totalFmt % t)

if __name__ == '__main__':
  main()

File: NodeJS_async/swapview.js
#!/usr/bin/env node

var sprintf = require('sprintf-js').sprintf;
var fs = require("fs");
var async = require("async");

function filesize(size){
	var units = "KMGT";
	var left = Math.abs(size);
	var unit = -1;
	for(;left > 1100 && unit < 3;unit++){
		left /= 1024;
	}
	if(unit==-1){
		return sprintf("%dB", size);
	}else{
		if(size<0) left= -left;
		return sprintf('%.1f%siB', left, units[unit]);
	}
}


function getSwapFor(pid, callback){
	async.map(["cmdline", "smaps"],
	function(name, c){
		fs.readFile("/proc/"+pid+"/"+name, encoding="utf-8", c);
	},
	function(err, results){
		if(err){ 
			callback(null, null);
			return; 
		}
		var comm = results[0];
                if(comm[comm.length-1] == '\0'){
                  comm = comm.substr(0, comm.length - 1)
                }
		comm = comm.replace(/\0/g, " ");
		var s=0.0;
		var smaps = results[1];
		smaps.split(/\n/).forEach(function (l){
			if(l.substr(0,5) == "Swap:"){
				s += parseInt(l.split(/[ ]+/)[1]);
			}
		});
		if(s>0){
			callback(null, [pid, s*1024, comm]);
		}else{
			callback(null, null);
		}
	});
}

function getSwap(callback){
	fs.readdir("/proc", function(err, data){
		async.map(data, function(spid, callback){
			var pid = parseInt(spid);
			if(pid > 0){
				getSwapFor(pid, callback);
			}else{
				callback(null, null);
			}
		}, function(err, results) {
			results = results.filter(function (x) {return x!=null;});
			results.sort(function (a, b){return a[1] - b[1];})
			callback(null, results);
		});
	});
}

getSwap(function(err, results){
	console.log(sprintf("%5s %9s %s", "PID", "SWAP", "COMMAND"));
	var t=0.0;
	results.forEach(function (s){
		console.log(sprintf("%5s %9s %s", s[0], filesize(s[1]), s[2]));
		t += s[1];
	});
	console.log(sprintf("Total: %8s", filesize(t)));
});
